src/core/src/type/nf4.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+// Contains logic derived from bitsandbytes
+// https://github.com/TimDettmers/bitsandbytes/blob/c82f51c0f784d8a43ebcb9cdefbf94e3f3b9c6c3/csrc/kernels.cu#L223
+// implementation.
+// Copyright notice from original source file is as follows.
+
+//*******************************************************************************
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+//==============================================================================
+
```

Comment:
please do not forget to add information about 3rd party code into 
https://github.com/openvinotoolkit/openvino/blob/master/licensing/third-party-programs.txt
and report it to OS PDT for the review.